### PR TITLE
Add command to fix score ordering

### PR DESCRIPTION
**Pull request by peppy**
Will be used to close https://github.com/ppy/osu-web/issues/10922.

I've started to test this in production but the elastic indexing (or something in between) isn't working well, even though the IDs are correctly updating.

The case I ran this on already if you'd like to check on it:

before:

| score id   | legacy score id |
|------------|-----------------|
| [1961647015](https://osu.ppy.sh/scores/1961647015) | [214615400](https://osu.ppy.sh/scores/fruits/214615400) |
| [1961647016](https://osu.ppy.sh/scores/1961647016) | [214584096](https://osu.ppy.sh/scores/fruits/214584096) |

after:

| score id   | legacy score id |
|------------|-----------------|
| [1961647015](https://osu.ppy.sh/scores/1961647015) | [214584096](https://osu.ppy.sh/scores/fruits/214584096) |
| [1961647016](https://osu.ppy.sh/scores/1961647016) | [214615400](https://osu.ppy.sh/scores/fruits/214615400) |

```
Verifying tied score orders for ruleset 2
Indexing to elasticsearch queue(s) osu-queue:scores_20240113
Processing 4157529...
4157529 has tied scores, checking order... FAIL
- legacy 214584096 remapping from 1961647016 to 1961647015
- legacy 214584099 remapping from 1961647024 to 1961647016
- legacy 214584117 remapping from 1961647032 to 1961647024
- legacy 214584216 remapping from 1961647049 to 1961647032
- legacy 214584353 remapping from 1961647057 to 1961647049
- legacy 214615400 remapping from 1961647015 to 1961647057
- legacy 214830862 remapping from 1961852270 to 1961852270
- legacy 216039456 remapping from 2422252508 to 2422252508
Reordering complete (1 reordered, 0/8821)
```

Will hold off doing a full run until reviewed.

### CODE REVIEW

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/ReorderIncorrectlyImportedTiedScoresCommand.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                    {
+                        ulong legacyScoreId = topScoresSorted[j].legacy_score_id!.Value;
+                        ulong oldScoreId = topScoresSorted[j].id;
+                        ulong newScoreId = topScores[j].id;
+
+                        Console.WriteLine($"- legacy {legacyScoreId} remapping from {oldScoreId} to {newScoreId}");
+
+                        if (!DryRun)
+                        {
+                            await conn.ExecuteAsync("UPDATE scores SET id = @newScoreId WHERE legacy_score_id = @legacyScoreId AND ruleset_id = @rulesetId", new
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/ReorderIncorrectlyImportedTiedScoresCommand.cs:105]: yep. it should bail if that happens though, at which point i'll rethink life.

**Comment by bdach** [comment]
I can't immediately see why the score indexer might be failing here *but* this might require special consideration for pinned scores. I haven't got a clue where that is right now, but if someone happens to have pinned one of these scores to their profile, this is gonna silently mess up things for them and make some other user's tied number 1 show on their profile.

**Review by bdach — COMMENTED** [review]
can't see much wrong with code that is already here, the above notwithstanding

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/ReorderIncorrectlyImportedTiedScoresCommand.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/ReorderIncorrectlyImportedTiedScoresCommand.cs:105]: Thinking about it there's _probably_ a case to be made here that `unix_updated_at` should also be updated? Although I'm not sure that realistically matters for preserved scores.

**Comment by peppy** [comment]
> but if someone happens to have pinned one of these scores to their profile, this is gonna silently mess up things for them and make some other user's tied number 1 show on their profile.

We've accounted* for this, see https://github.com/ppy/osu-web/pull/11003 (basically haven't migrated away from using old IDs just yet, note that PR is not deployed yet and won't be until i run this PR to completion).

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/ReorderIncorrectlyImportedTiedScoresCommand.cs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                {
+                    elasticQueueProcessor.PushToQueue(topScores.Select(s => new ElasticQueuePusher.ElasticScoreItem
+                    {
+                        ScoreId = (long)s.id
+                    }).ToList());
+                }
+
+                totalReordered++;
+
+                Console.WriteLine($"Reordering complete ({totalReordered} reordered, {i}/{beatmaps.Length})");
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/ReorderIncorrectlyImportedTiedScoresCommand.cs:127]: was intended to only count when one was actually reordered 😄

File: osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/ReorderIncorrectlyImportedTiedScoresCommand.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/ReorderIncorrectlyImportedTiedScoresCommand.cs:105]: I think it's fine to leave it.

**Comment by bdach** [comment]
> We've accounted* for this, see [ppy/osu-web#11003](https://github.com/ppy/osu-web/pull/11003) (basically haven't migrated away from using old IDs just yet, note that PR is not deployed yet and won't be until i run this PR to completion).

Um... I'm not sure that's enough?

Sure, old pins from before the migration to the new score structure will be fine. But I can activate lazer mode and pin a legacy score, but using the `solo_scores` pin type just fine, and if someone *has* done that, then applying this reordering operation can break the pin, no?

**Pull request merged**